packages/saved-views-react/src/useSavedViews.tsx

```diff
@@ -528,46 +530,71 @@ function createSavedViewActions(
         }
       },
     ),
-  };
-
-  /** Serializes action execution and notifies when action processing begins and ends. Silences cancellation errors. */
-  function actionWrapper<T extends unknown[]>(
-    callback: (...args: T) => Promise<void>,
-  ): (...args: T) => Promise<void> {
-    return async (...args) => {
-      ref.current.actionQueue.push(async () => {
-        if (signal.aborted) {
-          return;
-        }
+    uploadThumbnail: actionWrapper(
+      async (savedViewId: string, imageDataUrl: string) => {
+        let prevThumnbnail: ReactNode | string | undefined;
+        updateSavedView(savedViewId, (savedView) => {
+          prevThumnbnail = savedView.thumbnail;
+          savedView.thumbnail = imageDataUrl;
+        });
 
         try {
-          await callback(...args);
+          await client.uploadThumbnail({ savedViewId, image: imageDataUrl, signal });
         } catch (error) {
-          if (error && typeof error === "object" && "name" in error && error.name === "AbortError") {
-            // It's a cancellation error
-          } else {
-            throw error;
+          if (prevThumnbnail !== undefined) {
+            const restoredDisplayName = prevThumnbnail;
+            updateSavedView(savedViewId, (savedView) => {
+              savedView.thumbnail = restoredDisplayName;
+            });
           }
+          throw error;
         }
-      });
+      },
+    ),
+  };
 
-      // If there are no other queued actions, start executing the queue
-      if (ref.current.actionQueue.length === 1) {
-        onUpdateInProgress();
+  /**
+   * Serializes action execution and notifies when action processing begins and ends. Continues to execute actions
+   * after cancellation.
+   */
+  function actionWrapper<T extends unknown[], U>(callback: (...args: T) => Promise<U>): (...args: T) => Promise<U> {
```

Comment:
This is a cool function :D

Reply:
Thanks, it does what it's supposed to but I dislike that it exists. I'm considering dropping optimistic update feature to simplify things.